File: agent-app-go/internal/sql/schema_prompt_formatter.go
package sql

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// SchemaPromptFormatter ports SchemaPromptFormatter.kt 1:1: turns db://schema JSON into
// a compact TABLE/FOREIGN KEYS/KNOWN COLUMN VALUES/RULES textual block.
type SchemaPromptFormatter struct{}

func (f *SchemaPromptFormatter) Format(rawSchema string) string {
	var root map[string]any
	if err := json.Unmarshal([]byte(rawSchema), &root); err != nil {
		return rawSchema
	}

	tablesRaw, ok := root["tables"].(map[string]any)
	if !ok {
		return rawSchema
	}

	var sb strings.Builder
	sb.WriteString("DATABASE SCHEMA\n")

	tableNames := make([]string, 0, len(tablesRaw))
	for k := range tablesRaw {
		tableNames = append(tableNames, k)
	}
	sort.Strings(tableNames)

	for _, table := range tableNames {
		columns, _ := tablesRaw[table].([]any)
		colStrs := make([]string, len(columns))
		for i, c := range columns {
			colStrs[i] = fmt.Sprintf("%v", c)
		}
		sb.WriteString("TABLE " + table + " (" + strings.Join(colStrs, ", ") + ")\n")
	}

	if fks, ok := root["foreignKeys"].([]any); ok && len(fks) > 0 {
		sb.WriteString("FOREIGN KEYS\n")
		for _, fk := range fks {
			sb.WriteString(fmt.Sprintf("- %v\n", fk))
		}
	}

	if valueHints, ok := root["valueHints"].(map[string]any); ok && len(valueHints) > 0 {
		sb.WriteString("KNOWN COLUMN VALUES (SQL 식별자가 아닌 실제 데이터 값)\n")
		hintKeys := make([]string, 0, len(valueHints))
		for k := range valueHints {
			hintKeys = append(hintKeys, k)
		}
		sort.Strings(hintKeys)
		for _, column := range hintKeys {
			values, _ := valueHints[column].([]any)
			escaped := make([]string, len(values))
			for i, v := range values {
				escaped[i] = "'" + strings.ReplaceAll(fmt.Sprintf("%v", v), "'", "''") + "'"
			}
			sb.WriteString("- " + column + " = [" + strings.Join(escaped, ", ") + "]\n")
		}
	}

	sb.WriteString("RULES\n")
	sb.WriteString("- TABLE에 선언된 테이블과 컬럼만 SQL 식별자로 사용한다.\n")
	sb.WriteString("- KNOWN COLUMN VALUES의 값은 해당 컬럼의 비교값으로만 사용한다.\n")
	sb.WriteString("- JOIN은 FOREIGN KEYS에 선언된 관계만 사용한다.")

	return sb.String()
}

File: agent-app-ktor/src/main/kotlin/com/riwonace/agentktor/sql/SchemaPromptFormatter.kt
package com.riwonace.agentktor.sql

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory

/** MCP의 기계 판독용 스키마 JSON을 소형 모델이 오인하지 않는 SQL 지향 표현으로 바꾼다. */
class SchemaPromptFormatter(private val mapper: ObjectMapper) {

    private val log = LoggerFactory.getLogger(javaClass)

    /** 원본 JSON의 테이블·외래키·값 힌트를 서로 다른 의미 영역으로 명시한다. */
    fun format(rawSchema: String): String =
        try {
            val root = mapper.readTree(rawSchema)
            val tables = root.path("tables")
            require(tables.isObject) { "tables가 JSON 객체가 아닙니다." }

            buildString {
                appendLine("DATABASE SCHEMA")
                tables.properties().forEach { (table, columns) ->
                    append("TABLE ").append(table).append(" (")
                    append(columns.joinToString(", ") { it.asText() })
                    appendLine(")")
                }

                val foreignKeys = root.path("foreignKeys")
                if (foreignKeys.isArray && !foreignKeys.isEmpty) {
                    appendLine("FOREIGN KEYS")
                    foreignKeys.forEach { appendLine("- ${it.asText()}") }
                }

                val valueHints = root.path("valueHints")
                if (valueHints.isObject && !valueHints.isEmpty) {
                    appendLine("KNOWN COLUMN VALUES (SQL 식별자가 아닌 실제 데이터 값)")
                    valueHints.properties().forEach { (column, values) ->
                        append("- ").append(column).append(" = [")
                        append(values.joinToString(", ") { "'${escapeLiteral(it.asText())}'" })
                        appendLine("]")
                    }
                }

                appendLine("RULES")
                appendLine("- TABLE에 선언된 테이블과 컬럼만 SQL 식별자로 사용한다.")
                appendLine("- KNOWN COLUMN VALUES의 값은 해당 컬럼의 비교값으로만 사용한다.")
                append("- JOIN은 FOREIGN KEYS에 선언된 관계만 사용한다.")
            }
        } catch (e: Exception) {
            // 구버전 또는 다른 MCP 구현의 스키마도 계속 사용할 수 있도록 원문으로 폴백한다.
            log.warn("스키마 프롬프트 구조화 실패, 원본 사용: {}", e.message)
            rawSchema
        }

    private fun escapeLiteral(value: String): String = value.replace("'", "''")
}
